security/vault.go

```diff
@@ -0,0 +1,242 @@
+package security
+
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/gaia-pipeline/gaia"
+)
+
+const (
+	vaultName = ".gaia_vault"
+)
+
+// Vault is a secret storage for data that gaia needs to store encrypted.
+type Vault struct {
+	Path string
+	Cert []byte
+	data map[string][]byte
+	sync.RWMutex
+}
+
+// NewVault creates a vault which is a simple k/v storage medium with AES encryption.
+// The format is:
+// KEY=VALUE
+// KEY2=VALUE2
+func NewVault() (*Vault, error) {
+	v := new(Vault)
+	// Creating vault file
+	vaultPath := filepath.Join(gaia.Cfg.VaultPath, vaultName)
+	if _, osErr := os.Stat(vaultPath); os.IsNotExist(osErr) {
+		gaia.Cfg.Logger.Info("vault file doesn't exist. creating...")
+		_, err := os.Create(vaultPath)
+		if err != nil {
+			gaia.Cfg.Logger.Error("failed creating vault file: ", err.Error())
+			return nil, err
+		}
+	}
+
+	// Setting up certificate key content
+	c, err := InitCA()
+	if err != nil {
+		return nil, err
+	}
+	data, err := ioutil.ReadFile(c.caKeyPath)
+	if err != nil {
+		return nil, err
+	}
+	v.Cert = data
+	v.Path = vaultPath
+	v.data = make(map[string][]byte, 0)
+	return v, nil
+}
+
+// LoadSecrets decrypts the contents of the vault and fills up a map of data to work with.
+func (v *Vault) LoadSecrets() error {
+	r, err := ioutil.ReadFile(v.Path)
+	if err != nil {
+		return err
+	}
+	data, err := v.decrypt(r)
+	if err != nil {
+		return err
+	}
+	return v.parseToMap(data)
+}
+
+// SaveSecrets encrypts data passed to the vault in a k/v format and saves it to the vault file.
+func (v *Vault) SaveSecrets() error {
+	// open f
+	data := v.parseFromMap()
+	encryptedData, err := v.encrypt(data)
+	if err != nil {
+		return err
+	}
+	err = ioutil.WriteFile(v.Path, []byte(encryptedData), 0400)
+	return err
+}
+
+// GetAll returns all keys and values in a copy of the internal data.
+func (v *Vault) GetAll() []string {
+	m := make([]string, 0)
+	for k := range v.data {
```

Comment:
What happens when someone accesses `Add` and in the same time `GetAll`? Isn't this a race condition?

Reply:
Yes, there should be a Read Lock here. :/ Thanks.